api/types/webauthn/webauthn.proto

```diff
@@ -0,0 +1,141 @@
+syntax = "proto3";
+
+// Package WebAuthn maps WebAuthn messages to protocol buffers.
+//
+// The mapping is designed to match both the WebAuthn specification and the
+// capabilities of current browser implementations.
+//
+// REST-based Teleport APIs will make an effort to transmit or embed JSON
+// messages matching the github.com/duo-labs/webauthn reference implementation,
+// to allow for easy browser integration. gRPC APIs are not meant for REST use
+// and thus make no such promises, although the correspondence should be
+// obvious.
+//
+// Note that, ordinarily, various fields in WebAuthn messages are encoded using
+// "RawURLEncoding" (aka, base64 URL encoding without padding). This is not the
+// case for _any_ of the fields mapped here, all bytes fields are transmitted
+// raw/unencoded.
+package webauthn;
+
+import "gogoproto/gogo.proto";
+
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// -----------------------------------------------------------------------------
+// WebAuthn messages used by server storage.
+// -----------------------------------------------------------------------------
+
+// SessionData stored by the Relying Party during authentication ceremonies.
+// Mirrors https://pkg.go.dev/github.com/duo-labs/webauthn/webauthn#SessionData.
+message SessionData {
+    // Raw challenge used for the ceremony.
+    bytes challenge = 1 [ (gogoproto.jsontag) = "challenge,omitempty" ];
+    // Raw User ID.
+    bytes user_id = 2 [ (gogoproto.jsontag) = "userId,omitempty" ];
+    // Raw Credential IDs of the credentials allowed for the ceremony.
+    repeated bytes allow_credentials = 3 [ (gogoproto.jsontag) = "allowCredentials,omitempty" ];
+    // Requested user verification.
+    // https://www.w3.org/TR/webauthn-2/#enum-userVerificationRequirement.
+    string user_verification = 4 [ (gogoproto.jsontag) = "userVerification,omitempty" ];
+}
+
+// -----------------------------------------------------------------------------
+// Assertion (aka login).
+// -----------------------------------------------------------------------------
+
+// Credential assertion used for login ceremonies.
+message CredentialAssertion { PublicKeyCredentialRequestOptions public_key = 1; }
+
+// Request options necessary for credential assertions, aka login ceremonies.
+// See https://www.w3.org/TR/webauthn-2/#dictionary-assertion-options or
+// refer to navigator.credentials.get in your browser.
+message PublicKeyCredentialRequestOptions {
+    // Raw challenge used for assertion.
+    bytes challenge = 1;
+    // Timeout in milliseconds.
+    int64 timeout_ms = 2;
+    // Relying Party ID.
+    string rp_id = 3;
```

Comment:
Most of our existing messages today (and even the W3 spec linked above) are using `CamelCase` for field names, not `snake_case` - should we be consistent here?

Reply:
Good question. Proto style[1] is lower_snake_case, but it seems I'm the odd one out. Will update tomorrow.

[1] https://developers.google.com/protocol-buffers/docs/style

Reply:
Just had a chat, we decided to go ahead with proto style, given that it is a separate file.